subprojects/language-jvm/src/main/java/org/gradle/api/tasks/compile/AbstractCompile.java

```diff
@@ -124,4 +157,42 @@ public String getTargetCompatibility() {
     public void setTargetCompatibility(String targetCompatibility) {
         this.targetCompatibility = targetCompatibility;
     }
+
+    /**
+     * Convention to fall back to the 'destinationDir' output for backwards compatibility with plugins that extend AbstractCompile and override the deprecated methods.
+     */
+    @Deprecated
+    private class BackwardCompatibilityOutputDirectoryConvention extends AbstractReadOnlyProvider<Directory> {
```

Comment:
We should extract this somewhere so that it can be reused. Even better if it's just automatically wired in for `@ReplacedBy` properties.

Reply:
I'll add a follow up issue for this once this PR is merged.

Reply:
#11653